src/number/complete.rs

```diff
@@ -1517,24 +1520,35 @@ where
   Ok((i, (sign, integer, fraction, exp)))
 }
 
+/// Case-insensitive comparison of digits. Only works if `y` is only ASCII letters.
+#[inline]
+fn case_insensitive_cmp(x: &[u8], y: &[u8]) -> bool {
+    let d = (x.iter().zip(y.iter())).fold(0, |d, (xi, yi)| d | xi ^ yi);
+    // This uses the trick that 'a' - 'A' == 0x20, and this is true
+    // for all characters, so as long as `yi` is a valid ASCII letter,
+    // `xi ^ yi` can only be 0 or 0x20.
+    d == 0 || d == 32
+}
```

Comment:
why not write `d == 0 || d == 0x20` ? any source for this trick ?

Reply:
actually even better write `'a' - 'A'`

Reply:
or `matches(d, 0 | 'a' - 'A')`

Reply:
It's used pretty extensively, and sure, I'll modify that. But the logic is quite simple: it's a lot cheaper than `to_ascii_lowercase()`. You can see this [here](https://godbolt.org/z/nYMEKE7ao). Notice just how much more efficient the resulting ASM is:

```asm
example::fast_nocase_nan:
        cmp     rsi, 3
        jb      .LBB0_2
        mov     al, byte ptr [rdi]
        mov     cl, byte ptr [rdi + 1]
        xor     al, 78
        xor     cl, 65
        or      cl, al
        mov     al, byte ptr [rdi + 2]
        xor     al, 78
        or      al, cl
        test    al, -33
        sete    al
        ret
.LBB0_2:
        xor     eax, eax
        ret

example::slow_nocase_nan:
        cmp     rsi, 3
        jb      .LBB1_4
        mov     al, byte ptr [rdi]
        lea     ecx, [rax - 65]
        cmp     cl, 26
        setb    cl
        shl     cl, 5
        or      cl, al
        cmp     cl, 110
        jne     .LBB1_4
        mov     al, byte ptr [rdi + 1]
        lea     ecx, [rax - 65]
        cmp     cl, 26
        setb    cl
        shl     cl, 5
        or      cl, al
        cmp     cl, 97
        jne     .LBB1_4
        mov     al, byte ptr [rdi + 2]
        lea     ecx, [rax - 65]
        cmp     cl, 26
        setb    cl
        shl     cl, 5
        or      cl, al
        cmp     cl, 110
        sete    al
        ret
.LBB1_4:
        xor     eax, eax
        ret
```

The logic is quite simple: as long as one of the two strings is **all** letters, then `xi ^ yi` will only be `0x00` or `0x20` if  both are ASCII letters. Since there are no other bits that could possibly obfuscate the two (`0b100000` and `0b0`), folding with a bitwise or can only produce `0x20` or `0x00` if the result of `xi ^ yi` is `0x20` or `0x00`. In short, there's no cases where this can fail.

Reply:
> 
> 
> or `matches!(d, 0 | 'a' - 'A')`

Sure, but if we're already using magic numbers, we have to rely on **why** this works, which is why the comments document this. `0x20` sounds good though, since it matches exactly with the comment description.

Reply:
By the way maybe put this function in number module would avoid some code duplication

Reply:
> By the way maybe put this function in number module would avoid some code duplication

I was actually just about to ask that... Thanks for actually reading my mind ^_^

Reply:
Speaking of code deduplication @Stargateur, is it worth using macros since it seems like `recognize_float`, etc. all duplicate the exact same code just using `f32` and `f64`? It would only be local to that single module, and would not be exported.

Reply:
An other solution would be to use num, but meh. Gael is very good with rust macro so I think it's totally ok to use macro for that if you are good at it.

Reply:
FYI: `matches!()` is 1.42+, and we currently support 1.41+ (although we're free to increment that). I'm not sure this is worth breaking that, I'll do whatever more senior contributors think is better.

Reply:
not possible for nom 7 so would be breaking change for nom policy

Reply:
> not possible for nom 7 so would be breaking change for nom policy. There is a github action to check that anyway.

Ah ok, I thought Nom7 was just guaranteed to support 1.48+, but currently supports 1.41+.

Reply:
This has been moved to the module, and added better documentation for it:

https://github.com/Alexhuszagh/nom/blob/805e5be9d7e16a6fc0d86e3e6a16fd042d97fd3f/src/number/mod.rs#L17-L25